stream: implement WHATWG streams

jasnell opened the pull request:
This is an initial Work in progress *Experimental implementation* of the WHATWG streams standard.

The intent is for this to be a complete spec compliant implementation. This PR includes the web platform tests.

It is my intention to...

1. Use this as part of the QUIC/HTTP3 implementation, and eventually as part of a larger revamped `net` subsystem API.
2. Integrate this into the `fs/promises` `FileHandle` API.
3. Integrate this into the `Blob` API
4. Implement adapters and magic for existing node.js streams APIs

Once compiled, the module is available via `require('stream/web')`. This *does not* expose the globals, and I don't intend to expose the globals until after the implementation is ready to graduate from experimental.

For an example of how the QUIC API intends to use this... here's a snippet:

```js
session.onstream = async ({stream, respondWith}) => {
  console.log(await stream.readable.getReader().read());  // reads a single chunk

  respondWith({ body: getReadableStreamFromSomewhere() });
};
```

TODOS:

* Necessary to land initial implementation
  * [x] Finish implementation
    * [x] ReadableStream
    * [x] WritableStream
    * [x] TransformStream
    * [x] Async iteration of `ReadableStream`
    * [x] `pipeTo` and `pipeThrough`
    * [x] `postMessage()` transfer of `ReadableStream`, `WritableStream`, and `TransformStream`
    * [x] Validate passing WPT tests
  * [x] Add appropriate WPT tests
  * [x] Finish initial documentation
  * [x] Split into multiple files
  * [x] Move into `lib/internal/webstreams`
* Necessary for coming out of experimental
  * [ ] Revisit error codes strategy
  * [ ] Performance profiling and optimization 
  * [x] Potentially split into separate files
  * [ ] Potentially refactor to simplify code
  * [ ] Implement Node.js streams adapters
    * [ ] `stream.Readable` <-> `ReadableStream` Adapters
    * [ ] `stream.Writable` <-> `WritableStream` Adapters
    * [ ] `stream.Transform` <-> `TransformStream` Adapters
    * [ ] In `pipeTo` / `pipeThrough`, if the destination and source are adapters of `stream.Readable` and `stream.Writable`, fallback to Node.js `pipeline()` instead if possible. If source and destination are both backed by native `StreamBase`, fallback to native pipeline.
    * [ ] `stream.pipeline` integration if possible.
    * [ ] Implement `FileHandle.prototype.readable` to get a `ReadableStream` for the `FileHandle`
* Additional
  * [ ] Implement `Blob.prototype.readable` to get a `ReadableStream` for the `Blob`
  * [ ] Implement QUIC `Stream.prototype.readable`

XadillaX commented:
A deadly simple way is just using Deno's. (just joking

legendecas commented:
> A deadly simple way is just using Deno's.

IIUC, like said in the OP, it is planned to be "working with existing `require('stream')` API through low level adapters and magic". I don't believe it's a good idea to "adopt" Deno's implementation as it is not designed to be work with Node.js stream API in the first place.

ljharb commented on lib/internal/streams/whatwg.js:
```diff
@@ -0,0 +1,3336 @@
+    reader,
+  } = stream[kState];
+  assert(reader[kState].readRequests.length);
+  const readRequest = reader[kState].readRequests.shift();
```
```suggestion
  const readRequest = ArrayPrototypeShift(reader[kState].readRequests);
```

ljharb commented on lib/internal/streams/whatwg.js:
```diff
@@ -0,0 +1,3336 @@
+    reader,
+  } = stream[kState];
+  assert(reader[kState].readIntoRequests.length);
+  const readIntoRequest = reader[kState].readIntoRequests.shift();
```
```suggestion
  const readIntoRequest = ArrayPrototypeShift(reader[kState].readIntoRequests);
```

ljharb commented on lib/internal/streams/whatwg.js:
```diff
@@ -0,0 +1,3336 @@
+      headOfQueue.byteOffset,
+      bytesToCopy);
+    if (headOfQueue.byteLength === bytesToCopy) {
+      queue.shift();
```
```suggestion
      ArrayPrototypeShift(queue);
```

ljharb commented on lib/internal/streams/whatwg.js:
```diff
@@ -0,0 +1,3336 @@
+
+function readableByteStreamControllerShiftPendingPullInto(controller) {
+  assert(controller[kState].byobRequest === null);
+  return controller[kState].pendingPullIntos.shift();
```
```suggestion
  return ArrayPrototypeShift(controller[kState].pendingPullIntos);
```

ljharb commented on lib/internal/streams/whatwg.js:
```diff
@@ -0,0 +1,3336 @@
+      buffer,
+      byteOffset,
+      byteLength,
+    } = queue.shift();
```
```suggestion
    } = ArrayPrototypeShift(queue);
```

ljharb commented on lib/internal/streams/whatwg.js:
```diff
@@ -0,0 +1,3336 @@
+function writableStreamMarkFirstWriteRequestInFlight(stream) {
+  assert(stream[kState].inFlightWriteRequest.promise === undefined);
+  assert(stream[kState].writeRequests.length);
+  const writeRequest = stream[kState].writeRequests.shift();
```
```suggestion
  const writeRequest = ArrayPrototypeShift(stream[kState].writeRequests);
```

ljharb commented on lib/internal/streams/_whatwg.js:
```diff
@@ -0,0 +1,1195 @@
+    function shiftPending() {
+      if (this[kState].byobRequest !== null)
+        throw new ERR_INVALID_STATE('Pending read request');
+      return pendingPullIntos.shift();
```
```suggestion
      return ArrayPrototypeShift(pendingPullIntos);
```

ljharb commented on lib/internal/streams/_whatwg.js:
```diff
@@ -0,0 +1,1195 @@
+        for (let n = 0; n < bytesToCopy; n++)
+          buffer[destStart + n] = head[head.byteOffset + n];
+        if (head.byteLength === bytesToCopy) {
+          state.queue.shift();
```
```suggestion
          ArrayPrototypeShift(state.queue);
```

ljharb commented on lib/internal/streams/whatwg.js:
```diff
@@ -0,0 +1,3336 @@
+  const {
+    value,
+    size,
+  } = controller[kState].queue.shift();
```
```suggestion
  } = ArrayPrototypeShift(controller[kState]);
```

ljharb commented on lib/internal/streams/_whatwg.js:
```diff
@@ -0,0 +1,1195 @@
+          throw new ERR_INVALID_RETURN_VALUE.RangeError(
+            'a finite number', 'size', chunkSize);
+        }
+        state.queue.push({ value: chunk, size: chunkSize });
```
```suggestion
        ArrayPrototypePush(state.queue, { value: chunk, size: chunkSize }));
```

ljharb commented on lib/internal/streams/_whatwg.js:
```diff
@@ -0,0 +1,1195 @@
+        req.error(stream[kState].storedError);
+        break;
+      default:
+        readRequests.push(req);
```
```suggestion
        ArrayPrototypePush(readRequests, req);
```

ljharb commented on lib/internal/streams/_whatwg.js:
```diff
@@ -0,0 +1,1195 @@
+            type: 'byob',
+          };
+          if (controller[kState].pendingPullIntos.length) {
+            controller[kState].pendingPullIntos.push(desc);
+            this[kState].readIntoRequests.push(req);
```
```suggestion
            ArrayPrototypePush(controller[kState].pendingPullIntos, desc);
            ArrayPrototypePush(this[kState].readIntoRequests, req);
```

ljharb commented on lib/internal/streams/_whatwg.js:
```diff
@@ -0,0 +1,1195 @@
+          if (controller[kState].queueTotalSize) {
+
+          }
+          controller[kState].pendingPullIntos.push(desc);
+          this[kState].readIntoRequests.push(req);
```
```suggestion
          ArrayPrototypePush(controller[kState].pendingPullIntos, desc);
          ArrayPrototypePush(this[kState].readIntoRequests, req);
```

ljharb commented on lib/internal/streams/whatwg.js:
```diff
@@ -0,0 +1,3336 @@
+      size === Infinity) {
+    throw new ERR_INVALID_ARG_VALUE.RangeError('size', size);
+  }
+  controller[kState].queue.push({ value, size });
```
```suggestion
  ArrayPrototypePush(controller[kState].queue, { value, size });
```

ljharb commented on lib/internal/streams/whatwg.js:
```diff
@@ -0,0 +1,3336 @@
+    const actions = [];
+    if (!preventAbort) {
+      if (dest[kState].state === 'writable')
+        actions.push(writableStreamAbort(dest, error));
+      else
+        actions.push(PromiseResolve());
+    }
+    if (!preventCancel) {
+      if (source[kState].state === 'readable')
+        actions.push(readableStreamCancel(source, error));
+      else
+        actions.push(PromiseResolve());
```
```suggestion
        ArrayPrototypePush(actions, writableStreamAbort(dest, error));
      else
        ArrayPrototypePush(actions, PromiseResolve());
    }
    if (!preventCancel) {
      if (source[kState].state === 'readable')
        ArrayPrototypePush(actions, readableStreamCancel(source, error));
      else
        ArrayPrototypePush(actions, PromiseResolve());
```

ljharb commented on lib/internal/streams/whatwg.js:
```diff
@@ -0,0 +1,3336 @@
+function readableStreamAddReadRequest(stream, readRequest) {
+  assert(readableStreamHasDefaultReader(stream));
+  assert(stream[kState].state === 'readable');
+  stream[kState].reader[kState].readRequests.push(readRequest);
+}
+
+function readableStreamAddReadIntoRequest(stream, readIntoRequest) {
+  assert(readableStreamHasBYOBReader(stream));
+  assert(stream[kState].state !== 'errored');
+  stream[kState].reader[kState].readIntoRequests.push(readIntoRequest);
```
```suggestion
  ArrayPrototypePush(stream[kState].reader[kState].readRequests, readRequest);
}

function readableStreamAddReadIntoRequest(stream, readIntoRequest) {
  assert(readableStreamHasBYOBReader(stream));
  assert(stream[kState].state !== 'errored');
  ArrayPrototypePush(stream[kState].reader[kState].readIntoRequests, readIntoRequest);
```

ljharb commented on lib/internal/streams/whatwg.js:
```diff
@@ -0,0 +1,3336 @@
+    type: 'byob',
+  };
+  if (pendingPullIntos.length) {
+    pendingPullIntos.push(desc);
```
```suggestion
    ArrayPrototypePush(pendingPullIntos, desc);
```

ljharb commented on lib/internal/streams/whatwg.js:
```diff
@@ -0,0 +1,3336 @@
+      return;
+    }
+  }
+  pendingPullIntos.push(desc);
```
```suggestion
  ArrayPrototypePush(pendingPullIntos, desc);
```

ljharb commented on lib/internal/streams/whatwg.js:
```diff
@@ -0,0 +1,3336 @@
+  buffer,
+  byteOffset,
+  byteLength) {
+  controller[kState].queue.push({
```
```suggestion
  ArrayPrototypePush(controller[kState].queue, {
```

ljharb commented on lib/internal/streams/whatwg.js:
```diff
@@ -0,0 +1,3336 @@
+  if (autoAllocateChunkSize !== undefined) {
+    try {
+      const buffer = new ArrayBuffer(autoAllocateChunkSize);
+      pendingPullIntos.push({
```
```suggestion
      ArrayPrototypePush(pendingPullIntos, {
```

ljharb commented on lib/internal/streams/whatwg.js:
```diff
@@ -0,0 +1,3336 @@
+    resolve,
+    reject,
+  } = createDeferredPromise();
+  stream[kState].writeRequests.push({
```
```suggestion
  ArrayPrototypePush(stream[kState].writeRequests, {
```

ljharb commented on lib/internal/streams/whatwg.js:
```diff
@@ -0,0 +1,3336 @@
+  autoAllocateChunkSize) {
+  assert(stream[kState].controller === undefined);
+  if (autoAllocateChunkSize !== undefined) {
+    assert(Number.isInteger(autoAllocateChunkSize));
```
```suggestion
    assert(NumberIsInteger(autoAllocateChunkSize));
```

ljharb commented on lib/internal/streams/whatwg.js:
```diff
@@ -0,0 +1,3336 @@
+  stream[kState].controller = controller;
+
+  PromisePrototypeThen(
+    PromiseResolve((async () => startAlgorithm())()),
```
```suggestion
    (async () => startAlgorithm())()),
```
there's no need to PromiseResolve the return of an async function

ljharb commented on lib/stream.js:
```diff
@@ -56,6 +56,14 @@ ObjectDefineProperty(Stream, 'promises', {
+  configurable: true,
+  enumerable: true,
+  get() {
+    return require('stream/whatwg');;
```
```suggestion
    return require('stream/whatwg');
```

ljharb commented:
It seems kind of odd to have the specifier include a standards body’s name.

jasnell commented:
> It seems kind of odd to have the specifier include a standards body’s name.

That's something we can worry about later once the basic implementation is done.. but... Alternative suggestions?

ljharb commented:
I agree the bikeshed can wait :-) just pointing it out.

As for suggestions: what are the long-term plans in node here? Are "node streams" going to eventually be deprecated/replaced by "whatwg streams"? Or will node likely have two streams implementations for awhile? What's notable about "whatwg streams" besides that they're what's in browsers? Would "web streams" make more sense?

jasnell commented:
Existing streams aren't going anywhere. At least not for the foreseeable future. I have no intent here to impact existing streams at all.

Perhaps ... `require('streams/standard')`?

ljharb commented:
Not stoked on implying that web specs are somehow “the” standard, but it’s probably better than “whatwg”, if “web” isn’t viable :-)

ronag commented:
I’d very much like to have a discussion about our overall strategy and plan regarding streams in node before landing this. I’m not against it just want to make sure we have a reasonable plan we can communicate.

jasnell commented:
> I’d very much like to have a discussion about our overall strategy and plan regarding streams in node before landing this

Ok, but I would ask that that discussion not happen in this PR. Can I ask you to open either an issue or discussion topic seeded with your thoughts?

ronag commented:
Any premium art performance figures relative to node streams?

Is stream.pipeline support in scope for this PR or is that future work?

jasnell commented:
> Any preliminary performance figures ...

No. Not yet. I won't be working on performance optimizations until I'm sure the implementation is correct per the spec.

> Is stream.pipeline support in scope for this PR ...

No, the adapters to existing streams API will come in separate PRs

jasnell commented:
@ljharb:
> Not stoked on implying that web specs are somehow “the” standard, but it’s probably better than “whatwg”, if “web” isn’t viable :-)

Ok, I switched it now to `require('stream/web')` :-)

jasnell commented:
The implementation is here largely complete. Next step for me will be to pull the relevant WPT's in with a separate commit and start making sure the implementation is correct. After that, I'll work on the documentation with examples.

ljharb commented on lib/internal/streams/whatwg.js:
```diff
@@ -0,0 +1,4233 @@
+
+  if (remainderSize) {
+    const end = desc.byteOffset + desc.bytesFilled;
+    const remainder = buffer.slice(end - remainderSize, remainderSize);
```
do we cache Buffer slice?

jasnell replied:
As far as I know, no.

ronag replied:
Might want to make a new buffer and copy the slice in order to avoid a long living reference to pool.

jasnell replied:
@ronag:
> Might want to make a new buffer and copy the slice in order to avoid a long living reference to pool.

There's no need. The buffer at this point is not connected to the pool.

ljharb commented on lib/internal/streams/whatwg.js:
```diff
@@ -0,0 +1,4233 @@
+          ],
+          view));
+    }
+    if (view.byteLength === 0 || view.buffer.byteLength === 0) {
```
```suggestion
    if (view.byteLength === 0 || ArrayBufferPrototypeByteLength(view.buffer) === 0) {
```
not 100% sure if getters are available in primordials too, but i hope so

also i'm not 100% sure if `view.buffer` is always an ArrayBuffer

jasnell replied:
I do not believe they are. 

If view is an ArrayBufferView then view.buffer should always be an ArrayBuffer.

ljharb replied:
`.buffer` is on DataView.prototype also; it seems ideal to be robust there as well, if possible

jasnell replied:
DataView is acceptable here per the spec. 

It is possible to override `view.buffer` using, for instance, `Object.defineProperty()`, which would cause this to fail. We do not currently cache the getters from these. We could make use of `Relect.get()` here. It adds a bit of complexity tho.

ljharb commented on lib/internal/streams/whatwg.js:
```diff
@@ -0,0 +1,4233 @@
+        ],
+        chunk);
+    }
+    if (chunk.byteLength === 0 || chunk.buffer.byteLength === 0) {
```
same here

ljharb commented on lib/internal/streams/whatwg.js:
```diff
@@ -0,0 +1,4233 @@
+      'ReadableByteStreamController', { });
+  }
+
+  get [SymbolToStringTag]() { return 'ReadableByteStreamController'; }
```
why use a non-brand-checking getter here rather than a data property on the prototype?

jasnell replied:
Either works. I was already looking at changing this.

ljharb commented on lib/internal/streams/whatwg.js:
```diff
@@ -0,0 +1,4233 @@
+    });
+  }
+
+  get [SymbolToStringTag]() { return 'TransformStreamDefaultController'; }
```
same question here

ljharb commented on lib/internal/streams/whatwg.js:
```diff
@@ -0,0 +1,4233 @@
+    };
+    this.size =
+      FunctionPrototypeBind(
+        ByteLengthQueuingStrategy.prototype.size,
```
should this be cached, in case someone's mutated the prototype function prior to construction?

jasnell replied:
This makes it easier/possible to use destructuring while retaining the `this`. It's not perfect but it works.

ljharb replied:
that's not what i mean, the bind is fine - it's that you're looking up `.size` on the mutable `ByteLengthQueuingStrategy.prototype` at runtime, which means that it could have been deleted or replaced.

jasnell replied:
I've changed it now to refer to this.size. Binding to a mutable version of the function is fine given the purpose here

ljharb commented on lib/internal/streams/whatwg.js:
```diff
@@ -0,0 +1,4233 @@
+  /**
+   * @type {QueuingStrategySize}
+   */
+  size(chunk) { return chunk?.byteLength | 0; }
```
does this not need to ensure it's an "array buffer view"?

jasnell replied:
Not per the spec. Will look into it tho

ljharb commented on lib/internal/streams/whatwg.js:
```diff
@@ -0,0 +1,4233 @@
+    };
+    this.size =
+      FunctionPrototypeBind(
+        CountQueuingStrategy.prototype.size,
```
also here (re caching)

ljharb commented on lib/internal/streams/whatwg.js:
```diff
@@ -0,0 +1,4233 @@
+
+function transformStreamDefaultControllerPerformTransform(controller, chunk) {
+  const transformPromise =
+    PromiseResolve(
+      (async () => controller[kState].transformAlgorithm(chunk, controller))());
```
```suggestion
    (async () => controller[kState].transformAlgorithm(chunk, controller))();
```

targos commented on doc/api/whatwg_streams.md:
```diff
@@ -0,0 +1,666 @@
+# WHATWG Streams
```
```suggestion
# Web Streams API
```

targos commented on doc/api/index.md:
```diff
@@ -64,6 +64,7 @@
 * [VM](vm.md)
 * [WASI](wasi.md)
 * [Web Crypto API](webcrypto.md)
+* [Web Streams](whatwg_streams.md)
```
```suggestion
* [Web Streams API](web_streams.md)
```

Or `webstreams.md` to be consistent with `webcrypto.md`

targos commented on lib/internal/streams/whatwg.js:
```diff
@@ -0,0 +1,4283 @@
+'use strict';
```
rename to `web.js` ?

targos replied:
It's also a very long file. Could it be split into several ones?

jasnell replied:
Eventually, perhaps. While I'm still working on ensuring that the implementation is correct, keeping everything in a single file is the most convenient. Once I get into refactoring for performance optimization I'll look at splitting things up.

targos commented on lib/internal/streams/whatwg.js:
```diff
@@ -0,0 +1,4283 @@
+'use strict';
+
+/* eslint-disable no-use-before-define */
```
What is used before it's defined?

jasnell replied:
A couple of the classes. I get into refactoring for performance optimization I'll be seeing what can be moved around.

ljharb commented on lib/internal/streams/web.js:
```diff
@@ -0,0 +1,4394 @@
+    stream,
+  } = writer[kState];
+  switch (stream[kState].state) {
+    case 'erroror':
```
```suggestion
    case 'error':
```

ljharb commented on lib/internal/streams/web.js:
```diff
@@ -0,0 +1,4394 @@
+// the buffer, byteLength, and byteOffset properties on
+// ArrayBuffer and ArrayBufferView's may have been tampered with.
+
+function ArrayBufferViewGetBuffer(view) {
+  return ReflectGet(view.constructor.prototype, 'buffer', view);
+}
+
+function ArrayBufferViewGetByteLength(view) {
+  return ReflectGet(view.constructor.prototype, 'byteLength', view);
+}
+
+function ArrayBufferViewGetByteOffset(view) {
+  return ReflectGet(view.constructor.prototype, 'byteOffset', view);
+}
```
unfortunately, the `constructor` property is never reliable :-/ these might need to use internal brand-checking methods to determine the appropriate getter to use.

jasnell replied:
Yeah, for now I'm going to leave these as is and focus on the other bits. We can fix this piece up before it lands (hopefully) or before it progresses out of experimental

ljharb commented on lib/internal/streams/web.js:
```diff
@@ -0,0 +1,4394 @@
+      controller,
+      remainder,
+      0,
+      remainder.byteLength);
```
```suggestion
      ArrayBufferGetByteLength(remainder);
```

ljharb commented on lib/internal/streams/web.js:
```diff
@@ -0,0 +1,4394 @@
+  let elementSize = 1;
+  let ctor = DataViewCtor;
+  if (isArrayBufferView(view) && !isDataView(view)) {
+    elementSize = view.constructor.BYTES_PER_ELEMENT;
```
this is also unreliable due to `.constructor` as well as the constructor being mutable

ronag commented on lib/internal/streams/web.js:
```diff
@@ -0,0 +1,4433 @@
+
+      [SymbolAsyncIterator]() { return this; }
+    }, AsyncIteratorPrototype);
+  }
```
Rather than creating a custom iterator can't you just implement this as an async generator? Similar to what we do in node streams?

MattiasBuelens replied:
That could work. The async generator machinery takes care of queuing multiple calls to `next()` or `return()` through `AsyncGeneratorEnqueue` ([example](https://tc39.es/ecma262/#sec-asyncgenerator-prototype-next)), so you'd get that part of the required Web IDL semantics for free. However, you'd have to be careful to match the timings of all the promise resolutions and side effects, e.g. ensuring that the reader's lock is released **before** you yield `{ done: true }` or throw an error.

jasnell replied:
For now, the current implementation works. We can look at the possibility of switching to an async generator later.

MattiasBuelens reviewed (commented):
Very excited to see this! 🤩

Jamesernator commented:
> * Implement `Blob.prototype.readable` to get a `ReadableStream` for the `Blob`

Assuming `buffer.Blob` is meant to be compatible with the browser's version, the spec [already has a method](https://www.w3.org/TR/FileAPI/#stream-method-algo) for getting a `ReadableStream` from a Blob. All browsers now support this.

> * [ ]  Implement `FileHandle.prototype.readable` to get a `ReadableStream` for the `FileHandle`

A corresponding `FileHandle.prototype.writable` would be nice to for piping data into files.

MattiasBuelens commented on lib/internal/streams/web.js:
```diff
@@ -0,0 +1,4489 @@
+        // If this is the first read, delay by one microtask
+        // to ensure that the controller has had an opportunity
+        // to properly start and perform the initial pull.
+        // TODO(@jasnell): The spec doesn't call this out so
+        // need to investigate if it's a bug in our impl or
+        // the spec.
```
Interesting. I ran into [a similar oddity](https://github.com/MattiasBuelens/web-streams-polyfill/commit/bcc1a1b01133658bcd4554a3b46d4bfd47281846), but I never investigated further. 🤔

jasnell replied:
Yeah I'd spotted that in your impl. Once I'm passing the rest of the WPTs I'll be revisiting.

MattiasBuelens commented on lib/internal/streams/web.js:
```diff
@@ -0,0 +1,4489 @@
+    return readableStreamTee(this, false);
+  }
+
+  // The values() method is not defined in the streams standard
+  // spec but it is called out in the section on async iteration.
+  // So... that's weird.
```
The `values()` method comes from `async iterable<any>` on the `ReadableStream` Web IDL definition (step 5 of ["define the asynchronous iteration methods"](https://heycam.github.io/webidl/#define-the-asynchronous-iteration-methods)). 😉

jasnell replied:
oh! right... I remember that now!

Btw, your typescript reference implementation has been absolutely invaluable as a reference point while going through the WPTs :-)

addaleax commented on src/node_buffer.cc:
```diff
@@ -1158,6 +1159,60 @@ void GetZeroFillToggle(const FunctionCallbackInfo<Value>& args) {
+  uint32_t source_offset = args[3].As<Uint32>()->Value();
+  size_t bytes_to_copy = args[4].As<Uint32>()->Value();
+
+  CHECK(destination->ByteLength() - destination_offset >= bytes_to_copy);
+  CHECK(source->ByteLength() - source_offset >= bytes_to_copy);
```
```suggestion
  CHECK_GE(destination->ByteLength() - destination_offset, bytes_to_copy);
  CHECK_GE(source->ByteLength() - source_offset, bytes_to_copy);
```

jasnell commented on lib/internal/streams/web.js:
```diff
@@ -0,0 +1,4433 @@
+        if (reader[kState].stream === undefined) {
+          return PromiseResolve();
+        }
+        assert(!reader[kState].readRequests.length);
```
Awesome. Yeah, I spotted that and went to your implementation to double check! :-)

jasnell commented:
@addaleax ... I'd like to ask your opinion on the `ERR_` code strategy for this PR. If you notice, I make extensive use of `ERR_INVALID_STATE` currently. I just don't know if we should define specific errors codes for all the various possible errors here or stay generic with it. I know you have opinions on this and I'd value your input :-)

jasnell commented:
@ljharb @addaleax @ronag @targos ... I've switched this to "ready to review".

WPT's should be passing (they are on my local dev machine, we'll see what happens in CI). There are a couple of excluded tests in stream.json that I will be investigating later.

ljharb commented on lib/internal/streams/web.js:
```diff
@@ -0,0 +1,4632 @@
+const kState = Symbol('kState');
+const kType = Symbol('kType');
+
+const AsyncIteratorPrototype =
+  ObjectGetPrototypeOf(
+    ObjectGetPrototypeOf(
+      async function* () {}).prototype);
```
this seems useful to move to primordials

jasnell replied:
Yeah, except that it doesn't seem to actually work the way we need it too. This is one that I'm going to be investigating further because the `AsyncIteratorPrototype` here because it's not quite correct.

jasnell replied:
Ok, figured it out. It's still not perfect given that the `AsyncIteratorPrototype` is itself susceptible to prototype polution but at least the brand check now works.

ljharb commented on lib/internal/streams/web.js:
```diff
@@ -0,0 +1,4632 @@
+      return readableStreamPipeTo(
+        this,
+        destination,
+        !!(preventClose),
+        !!(preventAbort),
+        !!(preventCancel),
```
are these extra parens a node convention?
```suggestion
        !!preventClose,
        !!preventAbort,
        !!preventCancel,
```

jasnell replied:
They're a James Convention ;-) No problem removing them.

ljharb commented on lib/internal/streams/web.js:
```diff
@@ -0,0 +1,4632 @@
+    let started = false;
+    let current;
+
+    function nextSteps() {
+      if (done)
+        return PromiseResolve({ done: true, value: undefined });
+
+      if (reader[kState].stream === undefined) {
+        return PromiseReject(
+          new ERR_INVALID_STATE.TypeError(
+            'The reader is not bound to a ReadableStream'));
+      }
```
```suggestion
    async function nextSteps() {
      if (done)
        return { done: true, value: undefined };

      if (reader[kState].stream === undefined) {
        throw new ERR_INVALID_STATE.TypeError(
            'The reader is not bound to a ReadableStream');
      }
```

jasnell replied:
I'd rather not switch this to an async function. Given that it creates a Promise explicitly that is driven by the read request object and returns that, wrapping this in an async function ends up allocating additional unnecessary Promises that just add cost.

ljharb replied:
v8 has lots of optimizations that prevent unnecessary promises from being created when using async/await.

jasnell replied:
Take this script:

```js
const { createHook } = require('async_hooks');
let count = 0;
const hook = createHook({
  init(id, type) {
    if (type === 'PROMISE')
      count++;
  }
});

hook.enable();

process.on('exit', () => console.log(count));

function foo() {
  const p = new Promise(() => {});
  return p;
}

async function foo2() {
  const p = new Promise(() => {});
  return p;
}

eval(`${process.argv[2]}()`)
```

Running foo() creates 1 Promise.
Running foo2() creates 3.

```
root@DESKTOP-5KK9VIR:~/node/node# ./node ../tmp/a foo
1
root@DESKTOP-5KK9VIR:~/node/node# ./node ../tmp/a foo2
3
root@DESKTOP-5KK9VIR:~/node/node#
```

jasnell replied:
The additional promises created here also change the timing of the microtask queue completions which throw off several of the web platform tests as currently written.

jasnell replied:
For now, I'm going to leave this as is, without converting it to an async function.

ljharb commented on lib/internal/streams/web.js:
```diff
@@ -0,0 +1,4632 @@
+      return promise.promise;
+    }
+
+    function returnSteps(value) {
+      if (done)
+        return PromiseResolve({ done: true, value });
+      done = true;
+
+      if (reader[kState].stream === undefined) {
+        return PromiseReject(
+          new ERR_INVALID_STATE.TypeError(
+            'The reader is not bound to a ReadableStream'));
+      }
+      assert(!reader[kState].readRequests.length);
+      if (!preventCancel) {
+        const result = readableStreamReaderGenericCancel(reader, value);
+        readableStreamReaderGenericRelease(reader);
+        return PromisePrototypeThen(
+          result,
+          () => { return { done: true, value }; });
+      }
+
+      readableStreamReaderGenericRelease(reader);
+      return PromiseResolve({ done: true, value });
```
```suggestion
    async function returnSteps(value) {
      if (done)
        return { done: true, value };
      done = true;

      if (reader[kState].stream === undefined) {
        throw new ERR_INVALID_STATE.TypeError(
          'The reader is not bound to a ReadableStream');
      }
      assert(!reader[kState].readRequests.length);
      if (!preventCancel) {
        const result = readableStreamReaderGenericCancel(reader, value);
        readableStreamReaderGenericRelease(reader);
        await result;
        return { done: true, value };
      }

      readableStreamReaderGenericRelease(reader);
      return { done: true, value };
```

ljharb commented on lib/internal/streams/web.js:
```diff
@@ -0,0 +1,4632 @@
+    // TODO(@jasnell): Explore whether an async generator
+    // can be used here instead of a custom iterator object.
+    return ObjectSetPrototypeOf({
+      next() {
+        // If this is the first read, delay by one microtask
+        // to ensure that the controller has had an opportunity
+        // to properly start and perform the initial pull.
+        // TODO(@jasnell): The spec doesn't call this out so
+        // need to investigate if it's a bug in our impl or
+        // the spec.
```
maybe making this `async next() {` with an `await 0;` would give you the microtask more easily?

jasnell replied:
So.. fun story... changing this to an async function that uses an await breaks the WPT's because there is a test for checking that a patched `Promise.prototype.then` is not called and `await ...` uses the `Promise.prototype` ....

![image](https://user-images.githubusercontent.com/439929/123152659-a906d400-d419-11eb-956b-80bf467e00eb.png)

... so while, yes, using await here would be nicer, it's prone to prototype polution. Note... the few places I am using `await` in this code already do not trigger that WPT failure.

ljharb replied:
ah, that is a good point :-/

jasnell replied:
going to leave this as is, without converting to an async function

MattiasBuelens replied:
Hmm, that is unfortunate. Using `await` makes the implementation much more readable, but if it's not possible to avoid calling the (potentially patched) global `Promise.prototype.then` then it's a no-go. 😢 

Does other code in Node's `lib` use `await`, and does that code also risk running a potentially monkey-patched `.then`?

> Note... the few places I am using await in this code already do not trigger that WPT failure.

It looks like [the current WPT test with a patched `.then`](https://github.com/web-platform-tests/wpt/blob/master/streams/readable-streams/patched-global.any.js#L140-L141) immediately closes the writable end, and doesn't even wait for the `pipeTo()` promise to settle before completing the test (and restoring the original `.then`). Most likely, if we were to add a test that actually pushes a few chunks through the pipe while `.then` is patched, we'd have more failures. 😕

jasnell replied:
> Does other code in Node's lib use await, and does that code also risk running a potentially monkey-patched .then?

Yes. I'm not sure there's a great way around it other than moving entirely away from using async/await syntax in favor of only using the captured primordials.

Jamesernator replied:
> > Does other code in Node's lib use await, and does that code also risk running a potentially monkey-patched .then?
> 
> Yes. I'm not sure there's a great way around it other than moving entirely away from using async/await syntax in favor of only using the captured primordials.
> 

One could attach the original `Promise.prototype.then` as an own property and `await` would take it. Hard to say if v8 would recognise to optimise this away though.

ljharb commented on lib/internal/streams/web.js:
```diff
@@ -0,0 +1,4632 @@
+   *   done : boolean
+   * }>}
+   */
+  read() {
+    if (!isReadableStreamDefaultReader(this))
+      throw new ERR_INVALID_THIS('ReadableStreamDefaultReader');
+    if (this[kState].stream === undefined) {
+      return PromiseReject(
+        new ERR_INVALID_STATE.TypeError(
+          'The reader is not attached to a stream'));
+    }
```
```suggestion
  async read() {
    if (!isReadableStreamDefaultReader(this))
      throw new ERR_INVALID_THIS('ReadableStreamDefaultReader');
    if (this[kState].stream === undefined) {
      throw new ERR_INVALID_STATE.TypeError(
        'The reader is not attached to a stream');
    }
```

currently, this function sometimes throws a sync exception and sometimes returns a Promise, which is both zalgo and something the JS spec tries *very hard* not to do. The suggested change means that it only ever returns a promise and never synchronously throws.

MattiasBuelens replied:
It's also non-compliant to throw a sync exception here. 😉 Since `read()` returns a promise type, ["create an operation function"](https://heycam.github.io/webidl/#dfn-create-operation-function) requires that this type check from step 2.1.2.3:
> 3. If _esValue_ does not implement the interface _target_, throw a TypeError.

is caught and turned into a rejection:
> And then, if an exception _E_ was thrown:
> 1. If _op_ has a return type that is a promise type, then return ! Call(`%Promise.reject%`, `%Promise%`, «_E_»).

Same thing applies to `readable.cancel()`, `reader.cancel()`, `reader.closed`, `writable.abort()`, `writable.close()`, `writer.abort()`, `writer.close()`, `writer.write()` `writer.closed` and `writer.ready`. The IDL harness tests should cover all these cases.

ljharb commented on lib/internal/streams/web.js:
```diff
@@ -0,0 +1,4632 @@
+   * @param {any} reason
+   * @returns {Promise<void>}
+   */
+  cancel(reason) {
+    if (!isReadableStreamDefaultReader(this))
+      throw new ERR_INVALID_THIS('ReadableStreamDefaultReader');
+    if (this[kState].stream === undefined) {
+      return PromiseReject(new ERR_INVALID_STATE.TypeError(
+        'The reader is not attached to a stream'));
+    }
```
```suggestion
  async cancel(reason) {
    if (!isReadableStreamDefaultReader(this))
      throw new ERR_INVALID_THIS('ReadableStreamDefaultReader');
    if (this[kState].stream === undefined) {
      throw new ERR_INVALID_STATE.TypeError(
        'The reader is not attached to a stream');
    }
```
same as above.

there's a bunch of these, so i won't comment on the others - ideally they'd all be async, but if the web spec requires they not be, then that would be unfortunate.

ljharb commented on lib/internal/streams/web.js:
```diff
@@ -0,0 +1,4632 @@
+    // The highWaterMark value is not checked until the strategy
+    // is actually used, per the spec.
+    this[kState] = {
+      highWaterMark: Number(init.highWaterMark),
```
```suggestion
      highWaterMark: +init.highWaterMark,
```
this is identical but doesn't require calling the primordial

targos replied:
I think it more readable with Number()

ljharb replied:
I do agree, but this isn't an npm library, it's node, and syntax is better than calling a primordial. my guess is it's more performant as well.

ljharb commented on lib/internal/streams/web.js:
```diff
@@ -0,0 +1,4632 @@
+    }
+  }
+
+  async close() {
+    this[kState].port.postMessage({ type: 'close' });
+    this[kState].port.close();
+  }
```
nothing's awaited or returned here - so this will always be an immediately resolved promise for undefined. is this intentional?

ljharb commented on lib/internal/streams/web.js:
```diff
@@ -0,0 +1,4632 @@
+
+// ---- ReadableStream Implementation
+
+function readableStreamPipeTo(
```
should this also be an async function? it sometimes returns a rejected promise, and sometimes undefined.

MattiasBuelens replied:
That would seem safer, yes.

We can do the same with `ReadableStream.pipeTo()`, and remove the manual `try { } catch (error) { return PromiseReject(error) }` there.

jasnell replied:
So again, yes, but doing so breaks the web platform test that checks a patched Promise.prototype.then ...

ljharb commented on lib/internal/streams/web.js:
```diff
@@ -0,0 +1,4632 @@
+  return abort.promise;
+}
+
+function writableStreamClose(stream) {
+  const {
+    state,
+    writer,
+    backpressure,
+    controller,
+  } = stream[kState];
+  if (state === 'closed' || state === 'errored') {
+    return PromiseReject(
+      new ERR_INVALID_STATE.TypeError('WritableStream is closed'));
+  }
+  assert(state === 'writable' || state === 'erroring');
+  assert(!writableStreamCloseQueuedOrInFlight(stream));
+  stream[kState].closeRequest = createDeferredPromise();
+  const { promise } = stream[kState].closeRequest;
+  if (writer !== undefined && backpressure && state === 'writable')
+    writer[kState].ready.resolve?.();
+  writableStreamDefaultControllerClose(controller);
+  return promise;
+}
```
```suggestion
async function writableStreamClose(stream) {
  const {
    state,
    writer,
    backpressure,
    controller,
  } = stream[kState];
  if (state === 'closed' || state === 'errored') {
    throw new ERR_INVALID_STATE.TypeError('WritableStream is closed');
  }
  assert(state === 'writable' || state === 'erroring');
  assert(!writableStreamCloseQueuedOrInFlight(stream));
  stream[kState].closeRequest = createDeferredPromise();
  const { promise } = stream[kState].closeRequest;
  if (writer !== undefined && backpressure && state === 'writable')
    writer[kState].ready.resolve?.();
  writableStreamDefaultControllerClose(controller);
  return promise;
}
```

jasnell replied:
Changing this to an async function causes the web platform tests to fail in rather fun ways... 

![image](https://user-images.githubusercontent.com/439929/123158810-1702c980-d421-11eb-8da0-89ae7880e4c0.png)

I presume it is because of the additional Promises and microtasks that are created...

ljharb commented on lib/internal/streams/web.js:
```diff
@@ -0,0 +1,4632 @@
+  return transformStreamDefaultControllerPerformTransform(controller, chunk);
+}
+
+function transformStreamDefaultSinkAbortAlgorithm(stream, reason) {
+  transformStreamError(stream, reason);
+  return PromiseResolve();
+}
```
```suggestion
async function transformStreamDefaultSinkAbortAlgorithm(stream, reason) {
  transformStreamError(stream, reason);
}
```

ronag reviewed (commented):
Can we split lib/internal/streams/web.js  into multiple smaller files? I find it hard to review and work with such huge files.

jasnell commented:
> Can we split lib/internal/streams/web.js into multiple smaller files?

Now that this is passing all the WPT's, yeah for sure. Much easier working on the implementation all in one file :-) ... I'll split things out later today when I'm working on resolving the review comments already received.

mcollina commented:
can you put them in a different folder than `internal/streams`? Something like `internal/webstreaams`? I think this would be easier to maintain long term.

MattiasBuelens commented on lib/internal/streams/web.js:
```diff
@@ -0,0 +1,4636 @@
+    const viewBuffer = ArrayBufferViewGetBuffer(view);
+    const viewBufferByteLength = ArrayBufferGetByteLength(viewBuffer);
+
+    if (viewByteLength === 0 || !viewBufferByteLength === 0) {
```
```suggestion
    if (viewByteLength === 0 || viewBufferByteLength === 0) {
```

MattiasBuelens commented on lib/internal/streams/web.js:
```diff
@@ -0,0 +1,4643 @@
+        new ERR_INVALID_STATE.TypeError(
+          'View ArrayBuffer is zero-length or detached'));
+    }
+    // Supposed to assert here that the view's buffer is not
+    // detached, but there's no API available to use to check that.
```
Now that I think about it, if `viewBufferByteLength !== 0` then the view's buffer cannot be detached. So we don't actually need this assert in the specification. 😛

MattiasBuelens commented on lib/internal/streams/web.js:
```diff
@@ -0,0 +1,4643 @@
+    bytesFilled,
+  } = desc;
+
+  if (byteOffset + bytesFilled !== viewByteOffset)
+    throw new ERR_INVALID_ARG_VALUE.RangeError('view', view);
+
+  if (bytesFilled + viewByteOffset > byteLength)
+    throw new ERR_INVALID_ARG_VALUE.RangeError('view', view);
```
This step seems to be missing:
> 8. If firstDescriptor’s buffer byte length is not view.[[ViewedArrayBuffer]].[[ByteLength]], throw a RangeError exception.

It seems you don't have the "buffer byte length" field in your pull-into descriptor representation either. It's a fairly recent addition, and its value is only needed for this one specific check. But still, it's an important check. 😁

jasnell replied:
Ah right, I forgot to leave a todo comment here because I needed to check on this... The given view.byteLength does not have to match but it's underlying view.buffer.byteLength does? Is the assumption there that the view is most likely a slice of the byobRequest.view?

MattiasBuelens replied:
Correct.

The intention is that you call `.respondWithNewView()` with a view whose buffer is either `=== byobRequest.view.buffer`, or is the "transferred version" of that buffer. This is necessary to uphold the contract of `byobReader.read(view)`: that promise must be resolved with a `{ done, value }` tuple whose `value.buffer` is the transferred version of `view.buffer`. The consumer must get their entire buffer back, since it may contain data from previous reads that they still want to use later.

The view's byte offset must be the same as `byobRequest.view.byteOffset`. So exactly like in `.respond(bytesWritten)`, the newly written bytes must be placed at the start of the BYOB request view.

The view's byte length must be at most `byobRequest.view.byteLength`, but it can be shorter. Once again, this is the same as with `.respond(bytesWritten)`, there `bytesWritten` must also be `<= byobRequest.view.byteLength`.

Effectively, these two calls are equivalent:
```javascript
byobRequest.respond(bytesWritten);
byobRequest.respondWithNewView(byobRequest.view.subarray(0, bytesWritten));
```
The advantage of the latter is that you can pass a view that is backed by a transferred version of `byobRequest.view.buffer`. This opens up more advanced use cases, for example you could transfer the BYOB request to a worker to fill it, then transfer it back and call `.respondWithNewView()`. This isn't possible with `.respond()`, since that requires that `byobRequest.view` **must not** have been transferred.

See [whatwg/streams#1114 (comment)](https://github.com/whatwg/streams/pull/1114#discussion_r600081235) and [#1123](https://github.com/whatwg/streams/pull/1123) for more context.

jasnell replied:
Gotcha... makes perfect sense. Thank you! :-)

MattiasBuelens commented on doc/api/webstreams.md:
```diff
@@ -0,0 +1,1096 @@
+
+const SECOND = 1000;
+
+const stream = new ReadableStream({
+  async start(controller) {
+    for await (const _ of every(SECOND))
+      controller.enqueue(performance.now());
+  }
+});
```
Interesting! I was going to comment that this stream would run forever even when canceled, because it doesn't have a `cancel()` method that stops the `for await of` loop. But actually, this still works: after the stream is canceled, any subsequent calls to `controller.enqueue()` will error, so the next iteration will exit the loop and close the `every()` async iterator. 😅

jasnell replied:
Yeah I had to test it to make sure also and it works really well!

MattiasBuelens commented on doc/api/webstreams.md:
```diff
@@ -0,0 +1,1096 @@
+  });
+};
+
+port2.postMessage(stream);
```
To transfer a stream (or really *any* transferable object), you need to add it to the `transferList`.
```suggestion
port2.postMessage(stream, [stream]);
```

jasnell replied:
Good catch :-)

domenic replied:
Note that the modern easier-to-read syntax (not sure if Node supports it yet) is `postMessage(stream, { transfer: [stream] })`

jasnell commented:
@ronag @mcollina ... I've split the implementation up into multiple files and moved them into `lib/internal/webstreams`

mcollina reviewed (commented):
I would restrain from using Array primordials as they are known to be causing performance problems.

jasnell commented:
Investigating the relevant CI failures:

* [ ] [Linux One](https://ci.nodejs.org/job/node-test-commit-linuxone/27807/) -- readable-byte-streams/general.any.js ... multiple failures with expected content size mismatches
* [ ] [AIX](https://ci.nodejs.org/job/node-test-commit-aix/37255/) -- readable-byte-streams/general.any.js ... multiple failures with expected content size mismatches
* [ ] [Windows](https://ci.nodejs.org/job/node-test-commit-windows-fanned/43474/) -- build time failure

davidfekke commented:
Extremely cool James. Thanks for your contributions to Node.js!

ljharb commented on lib/internal/webstreams/readablestream.js:
```diff
@@ -0,0 +1,2689 @@
+   * @param {StreamPipeOptions} [options]
+   * @returns {Promise<void>}
+   */
+  pipeTo(destination, options = {}) {
```
this function sometimes throws synchronously, and sometimes returns a rejected promise.

i think all the throws need to become "return Promise.reject()"

jasnell replied:
The synchronous throws are all wrapped in a try catch that returns a rejected promise.

richardlau commented:
> Investigating the relevant CI failures:
> 
> * [ ]  [Windows](https://ci.nodejs.org/job/node-test-commit-windows-fanned/43474/) -- build time failure ---= @nodejs/build @nodejs/platform-windows could someone from the build/infrastructure team take a look here. Building on windows is working great for me locally but failing in CI. It looks specific to the CI platform.

From https://ci.nodejs.org/job/node-compile-windows/41038/nodes=win-vs2019/console
```
01:24:24   Traceback (most recent call last):
01:24:24     File "C:\workspace\node-compile-windows\node\tools\js2c.py", line 222, in <module>
01:24:24       main()
01:24:24     File "C:\workspace\node-compile-windows\node\tools\js2c.py", line 218, in main
01:24:24       JS2C(source_files, options.target)
01:24:24     File "C:\workspace\node-compile-windows\node\tools\js2c.py", line 142, in JS2C
01:24:24       AddModule(filename, definitions, initializers)
01:24:24     File "C:\workspace\node-compile-windows\node\tools\js2c.py", line 116, in AddModule
01:24:24       code = ReadFile(filename)
01:24:24     File "C:\workspace\node-compile-windows\node\tools\js2c.py", line 43, in ReadFile
01:24:24       with codecs.open(filename, "r", "utf-8") as f:
01:24:24     File "C:\Python39\lib\codecs.py", line 905, in open
01:24:24       file = builtins.open(filename, mode, buffering)
01:24:24   FileNotFoundError: [Errno 2] No such file or directory: 'deps/acorn-plugins/crn-piate-mehods/index.js'
```

@jasnell You're probably running into the same issue that @joyeecheung did in https://github.com/nodejs/node/pull/38971#issuecomment-863511329 that https://github.com/nodejs/node/pull/39069 looks to address.

ljharb commented on lib/internal/webstreams/writablestream.js:
```diff
@@ -0,0 +1,1270 @@
+  stream[kState].closeRequest = createDeferredPromise();
+  const { promise } = stream[kState].closeRequest;
+  if (writer !== undefined && backpressure && state === 'writable')
+    writer[kState].ready.resolve?.();
```
is this user code being invoked? if so, it needs to be try/catched so a rejected promise can be returned if needed

MattiasBuelens replied:
No, these are deferreds created in [`setupWritableStreamDefaultWriter`](https://github.com/nodejs/node/blob/721b8cdbee12fa6ad5700dcbc643bae733322f1b/lib/internal/webstreams/writablestream.js#L508). While `writer[kState].ready.promise` is exposed to author code as `writer.ready`, any fulfillment/rejection handlers attached to that promise are run in a separate microtask.

MattiasBuelens commented on test/wpt/status/streams.json:
```diff
@@ -16,5 +16,8 @@
     "fail": "Needs more investigation"
+  },
+  "readable-byte-streams/general.any.js": {
+    "fail": "Web Platform Tests fail to consider platform endianness"
```
Shoot, you're right. We're [writing to a `Uint8Array`](https://github.com/web-platform-tests/wpt/blob/a658b79cd0d686edf6e5eb718a0e55e9a3933e67/streams/readable-byte-streams/general.any.js#L940-L942) and [reading it back from a `Uint16Array`](https://github.com/web-platform-tests/wpt/blob/a658b79cd0d686edf6e5eb718a0e55e9a3933e67/streams/readable-byte-streams/general.any.js#L960), incorrectly assuming that we're on a little-endian platform. I think I spotted that a while ago, but I forgot to make an issue about it. I'll look into it.

jasnell replied:
I've at least manually verified that the implementation is doing the right thing here. I will revisit this after the WPT's are updated.

MattiasBuelens replied:
web-platform-tests/wpt#29488 has landed, you should now be able to re-enable these tests. 🤞

jasnell replied:
That was fast! Thank you!

jasnell commented:
@mcollina:
> I would restrain from using Array primordials as they are known to be causing performance problems.

Given that none of this has been performance optimized yet, I'd rather leave those in for now. Once we get to the performance optimization step we can look at changing those back to the non-primordial versions.

MattiasBuelens reviewed (commented):
I've managed to get the IDL harness tests to run using [a bit of a hack](https://github.com/nodejs/whatwg-stream/commit/02478519c5f11ad2e5864caaa94f79d77549a165), and I've been fixing some compliance issues. If you want, I can open a PR to your branch? 🙂

I also noticed that you're repeatedly squashing all changes into a single commit and force-pushing it to this branch. I find that this makes it difficult to see what has changed over time. Would it be possible to push the changes as separate commits instead? We can always squash them together right before this PR gets merged.

jasnell commented:
A PR to the branch is just fine! And welcome! On the IDL variances, I think we should skip those for now. The options here are limited and I'll need to look at the choices.

jasnell commented on lib/internal/webstreams/readablestream.js:
```diff
@@ -0,0 +1,2689 @@
+ *
+ */
+
+class ReadableStream extends JSTransferable {
```
Unfortunately not without another hack that breaks other tests. We're a bit limited here

jasnell replied:
To clarify, it's possible by dipping down into C++ land to create the transferable classes and use Object.defineOwnProperties to add the methods and accessors in JS land. It's not as elegant from a syntax and maintenance perspective but we can't clone/transfer objects unless the JS object is attached at the native layer to our BaseObject. The other way is to use the makeTransferable utility method we added recently but that breaks a couple of the other tests because it also messes with the prototype and constructor.

jasnell replied:
Since this PR is passing CI and it's already pretty complicated as it is, I'll make the changes we need to get the IDL harness test completely passing in a separate follow-on PR

jasnell commented on lib/internal/webstreams/readablestream.js:
```diff
@@ -0,0 +1,2689 @@
+  source,
+  highWaterMark,
+  sizeAlgorithm) {
+  const controller = new InternalReadableStreamDefaultController();
```
Will look at a better way of accomplishing this :-)

jasnell replied:
Ok, replaced the stand-in classes with uses of `Reflect.construct` that will ensure that the prototype and constructor are reported correctly. There's still the issue with the JSTransferable base, which I'll work on separately.

jasnell commented:
> Would it be possible to push the changes as separate commits instead?

From here out yeah. For draft branches I typically aggressively squash. I forgot that I switched this to ready! ;-)

Trott commented:
> > Would it be possible to push the changes as separate commits instead?
> 
> From here out yeah. For draft branches I typically aggressively squash. I forgot that I switched this to ready! ;-)

If you commit with `git commit --fixup HEAD` (or `--fixup <commit-hash>`), then the commit-queue bot will correctly squash them into the existing commit. `node-core-utils` will also honor them. Makes the rebasing step in the landing process easier and less error-prone, I think.)  (And if you already knew all that and are choosing to do it differently for whatever reason, then hey, this is a PSA for other contributors/collaborators.)

targos commented on test/wpt/status/streams.json:
```diff
@@ -0,0 +1,17 @@
+    "fail": "TODO: implement detached ArrayBuffer support"
+  },
+  "readable-byte-streams/non-transferable-buffers.any.js": {
+    "fail": "Test is buggy - https://github.com/web-platform-tests/wpt/issues/29500"
```
bug was fixed :)

targos commented:
Coverage: https://codecov.io/gh/nodejs/node/tree/511cfad6d819227e983ec0fbe290dd4e8b6fcdfa/lib/internal/webstreams

jasnell commented:
Added more test coverage! Still not 100% because there are a couple of things I'm not sure yet how to test and the WPT's don't cover them

MattiasBuelens commented on lib/internal/webstreams/readablestream.js:
```diff
@@ -1609,6 +1600,11 @@ function readableStreamFulfillReadRequest(stream, chunk, done) {
   assert(reader[kState].readRequests.length);
   const readRequest = ArrayPrototypeShift(reader[kState].readRequests);
+
+  // TODO(@jasnell): It's not clear under what exact conditions done
+  // will be true here. The spec requires this check but none of the
+  // WPT's or other tests trigger it. Will need to investigate how to
+  // get coverage for this.
```
The reference implementation also has no coverage on these lines... It looks like this might be dead code in the specification. 🤔 

There are 3 places where the spec calls `ReadableStreamFulfillReadRequest`. However, 2 of those are from `Enqueue()` and always pass `done = false`, so they don't count.

The remaining place is [step 5.1 in `ReadableByteStreamControllerCommitPullIntoDescriptor`](https://streams.spec.whatwg.org/commit-snapshots/d0bd912c2c1547aabee61c950ff624e0937d8779/#ref-for-readable-stream-fulfill-read-request%E2%91%A0). This is called from 3 places:
1. [Step 2.3.2 of `ReadableByteStreamControllerProcessPullIntoDescriptorsUsingQueue`](https://streams.spec.whatwg.org/commit-snapshots/d0bd912c2c1547aabee61c950ff624e0937d8779/#ref-for-readable-byte-stream-controller-commit-pull-into-descriptor). However, this operation starts by asserting that `closeRequested = false`, which implies `state = 'readable'`. None of the steps in this operation change `state`, so by the time we reach 2.3.2 it'll still be `'readable'` and `done` will be set to `false`.
2. [Step 3.1.2 of `ReadableByteStreamControllerRespondInClosedState`](https://streams.spec.whatwg.org/commit-snapshots/d0bd912c2c1547aabee61c950ff624e0937d8779/#ref-for-readable-byte-stream-controller-commit-pull-into-descriptor%E2%91%A0). However, step 3 checks if the stream has a BYOB reader attached, so we'll always hit step 6.2 (`ReadableStreamFulfillReadIntoRequest`) instead of step 5.1 (`ReadableStreamFulfillReadRequest`) of `ReadableByteStreamControllerCommitPullIntoDescriptor`.
3. [Step 8 of `ReadableByteStreamControllerRespondInReadableState`](https://streams.spec.whatwg.org/commit-snapshots/d0bd912c2c1547aabee61c950ff624e0937d8779/#ref-for-readable-byte-stream-controller-commit-pull-into-descriptor%E2%91%A1). However, this is once again only called when `state = 'readable'`, and none of the preceding steps can change `state` to `'closed'`.

Maybe we should change the spec and turn this into an `assert(done === false)` or something... 😛

jasnell replied:
Yeah I was thinking the same. I tried hard to find ways of triggering that path, even artificially, and couldn't ever hit it.

MattiasBuelens commented on test/parallel/test-whatwg-readablestream.js:
```diff
@@ -1290,3 +1312,211 @@ class Source {
 }
+
+{
+  assert.throws(() => Reflect.get(ReadableStream.prototype, 'locked', {}), {
+    code: 'ERR_INVALID_THIS',
+  });
```
These kinds of tests should also be covered by the IDL harness tests... but I take it we haven't yet managed to make those run correctly in Node? 😞

jasnell replied:
Once we get the IDL harness test working we can remove these. With your hack the IDL tests do work so maybe we can consider adding that to the wpt's?

MattiasBuelens replied:
I don't think we should upstream this to WPT as-is. It's *really* hacky, and it may not be true for other environments. Maybe someone else wants to implement a `Window`-exposed interface in their project, and run these same IDL tests against their implementation.

However, I found a better way. We can add this to the `runner.setInitScript()` call:
```javascript
// For IDL harness tests, pretend that we're a DedicatedWorker.
global.DedicatedWorkerGlobalScope = global.constructor;
```
This makes us pass this check in [idlharness.js](https://github.com/web-platform-tests/wpt/blob/9e11b662c7204cf561801314bec46ec36220ebc6/resources/idlharness.js#L656-L659):
```javascript
if ('DedicatedWorkerGlobalScope' in self &&
    self instanceof DedicatedWorkerGlobalScope) {
    return globals.has("DedicatedWorker");
}
```
This is still a hack, but it's *much* less invasive and more flexible:
* We don't need to patch `idlharness.js`, so we can stay in sync with upstream WPT.
* We can put this in the specific test script itself (i.e. `test-streams.js`), so other tests aren't affected. In fact, other tests could use a similar hack if they want to test interfaces exposed on `Window` or `SharedWorker` or whatever. 😁

Thoughts?

MattiasBuelens replied:
Hang on, that causes [`create_test_environment` in `testharness.js`](https://github.com/web-platform-tests/wpt/blob/master/resources/testharness.js#L475-L478) to create a `DedicatedWorkerTestEnvironment` instead of a `ShellTestEnvironment`. That environment uses `self.postMessage()` to send back the results back, so the tests never complete... Dangit. 😬

MattiasBuelens replied:
Well, this works. But now it's leaning more towards the hacky side. 😛 
```javascript
Object.defineProperties(global, {
  DedicatedWorkerGlobalScope: {
    get() {
      // Pretend that we're a DedicatedWorker, but *only* for the IDL harness.
      // For the regular test harness, keep the JavaScript shell environment.
      if (new Error().stack.includes("idlharness.js")) {
        // self instanceof DedicatedWorkerGlobalScope === true
        return global.constructor;
      } else {
        // self instanceof DedicatedWorkerGlobalScope === false
        return function() {};
      }
    }
  }
});
```
Alternatively, we add a special case in [`test/common/wpt/worker.js`](https://github.com/nodejs/node/blob/ccc50a982044812e9e307a4626442c564cd60cf8/test/common/wpt/worker.js#L54) to temporarily add this global right before we load `idlharness.js`. But then it's shared across all WPT tests again... 😕 Or we have to add API to `WPTRunner` to choose which environment to emulate for the IDL harness.

jasnell replied:
I think the defineProperties approach works fine for now. We can revisit this again later.

jasnell replied:
Ok, IDL test is now working with the defineProperties. I'm going to go ahead and keep the additional checks I added in our own tests, just because they also check the error code.

targos commented:
New coverage looks very good! https://codecov.io/gh/nodejs/node/tree/adc7160f83790698048aa4761aec623a2c722530/lib/internal/webstreams

mcollina reviewed (approved):
lgtm

jasnell commented:
@nodejs/build ... getting this on linux-containered (https://ci.nodejs.org/job/node-test-commit-linux-containered/27622/nodes=ubuntu1804_sharedlibs_debug_x64/testReport/junit/(root)/test/parallel_test_child_process_fork_exec_path/)

```
Error: ENOSPC: no space left on device, copyfile '/home/iojs/build/workspace/node-test-commit-linux-containered/out/Debug/node' -> '/home/iojs/tmp/.tmp.408/node-copy.exe'
    at Object.copyFileSync (node:fs:2790:3)
    at Object.<anonymous> (/home/iojs/build/workspace/node-test-commit-linux-containered/test/parallel/test-child-process-fork-exec-path.js:51:4)
    at Module._compile (node:internal/modules/cjs/loader:1095:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1124:10)
    at Module.load (node:internal/modules/cjs/loader:975:32)
    at Function.Module._load (node:internal/modules/cjs/loader:816:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:79:12)
    at node:internal/main/run_main_module:17:47 {
  errno: -28,
  syscall: 'copyfile',
  code: 'ENOSPC',
  path: '/home/iojs/build/workspace/node-test-commit-linux-containered/out/Debug/node',
  dest: '/home/iojs/tmp/.tmp.408/node-copy.exe'
}
```

jasnell commented:
Squashed the commits, rebased from the main, and CI passing. This is ready for reviews and landing. While it has the minimum number of sign offs necessary to land, I'll give it until Wednesday before landing.

targos reviewed (approved):
Rubberstamp LGTM as experimental.

jasnell commented:
Landed in a9f9d2b3b7ca...fa0c6883a657

jimmywarting commented:
> - [ ] Implement `FileHandle.prototype.readable` to get a ReadableStream for the FileHandle

I was like: Wait what? Dose node have something called a FileHandle must be something i have missed. But turns out that was not the same as [FileSystemFileHandle](https://wicg.github.io/file-system-access/#api-filesystemfilehandle) provided by Native file system access (which i totally think we should implement btw, Deno have considered this too)

I don't know how i feel about `FileHandle.prototype.readable` i think it should perhaps mimic more how the file system access works (you get a FileHandle, call `FileHandle.prototype.getFile()` and receive a Blob/File backed up by the filesystem (#37340) 

```js
const file = await fileHandle.getFile()
const stream = file.stream()
console.assert(file instancesof Blob) // true
console.assert(stream instancesof ReadableStream) // true
```

---


> - [ ] Implement `Blob.prototype.readable` to get a ReadableStream for the Blob

I hope you meant `Blob.prototype.stream()`?
